tests/kernel/smp/src/main.c

```diff
@@ -37,45 +52,354 @@ void t2_fn(void *a, void *b, void *c)
+		k_thread_abort(tdata[i].tid);
+	}
+	/* Test point: All THREADS_NUM spawned threads must have
+	 * executed.
+	 */
+	for (i = 0; i < THREADS_NUM; i++) {
+		zassert_true(tdata[i].executed == 1, "Threads didn't get"
+			     " scheduled on all the cores");
+	}
+}
+
+/**
+ * @brief Verify round robin on multiple cores with preempt threads
+ *
+ * @details The test basically validates for the scenario
+ * where parent has to execute after all child processes
+ * execution is complete
+ */
+static void test_smp_multi_core_preemp_rr_threads(void)
+{
+	int i;
+
+	k_thread_priority_set(k_current_get(),
+			      K_PRIO_PREEMPT(K_LOWEST_THREAD_PRIO));
+
+	k_sched_time_slice_set(100, 0);
+
+	for (int i = 0; i < THREADS_NUM; i++) {
+		tdata[i].executed = 0;
+		tdata[i].tid = k_thread_create(&tthread[i], tstack[i],
+					       STACK_SIZE, smp_thread_entry,
+					       (void *)i, NULL, NULL,
+					       K_PRIO_PREEMPT(
+						       K_LOWEST_THREAD_PRIO - 1),
+					       0, 0);
+	}
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		k_thread_abort(tdata[i].tid);
+	}
+	/* Test point: All THREADS_NUM spawned threads must have
+	 * executed.
+	 */
+	for (i = 0; i < THREADS_NUM; i++) {
+		zassert_true(tdata[i].executed == 1, "Threads didn't get"
+			     " scheduled on all the cores");
+	}
+}
+
+/**
+ * @brief Verify round robin on multi core with equal priority threads
+ *
+ * @details The test creates THREAD_NUM of preemptive equal
+ * priority threads with priority equal to parent thread.
+ * All threads including parent thread should run with round
+ * robin scheduling algorithm.
+ */
+static void test_smp_multi_core_preemp_equal_rr_threads(void)
+{
+	int i;
+
+	k_thread_priority_set(k_current_get(),
+			      K_PRIO_PREEMPT(K_HIGHEST_THREAD_PRIO));
+
+	k_sched_time_slice_set(100, 0);
+
+	for (int i = 0; i < THREADS_NUM; i++) {
+		tdata[i].executed = 0;
+		tdata[i].tid  = k_thread_create(&tthread[i], tstack[i],
+						STACK_SIZE, smp_thread_entry,
+						(void *)i, NULL, NULL,
+						K_PRIO_PREEMPT(
+							K_HIGHEST_THREAD_PRIO), 0, 0);
+	}
+
+	k_sleep(THREAD_DELAY * THREADS_NUM / CONFIG_MP_NUM_CPUS);
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		k_thread_abort(tdata[i].tid);
+	}
+	/* Test point: All THREADS_NUM spawned threads must have
+	 * executed.
+	 */
+	for (i = 0; i < THREADS_NUM; i++) {
+		zassert_true(tdata[i].executed == 1, "Threads"
+			     " didn't get scheduled on all the cores");
+	}
+}
+
+/**
+ * @brief Verify multi core processing of delayed threads
+ *
+ * @details Create THREAD_NUM threads with delayed start and
+ * check if they have executed
+ */
+static void test_smp_delayed_threads(void)
+{
+	int i;
+
+	k_sched_time_slice_set(100, 0);
+
+	for (int i = 0; i < THREADS_NUM; i++) {
+		tdata[i].executed = 0;
+		tdata[i].tid = k_thread_create(&tthread[i], tstack[i],
+					       STACK_SIZE, smp_thread_entry,
+					       (void *)i, NULL, NULL,
+					       K_PRIO_PREEMPT(
+						       K_HIGHEST_THREAD_PRIO), 0, 50);
```

Comment:
Would be a little more obvious to make this "thread_start_delay_ms" or something instead of "50".   I'm not a zealout about magic number or anything, but k_thread_create() has a zillion parameters and no one can remember which is which.